File: algorithms/2_medium/G/02017_grid_game/solution.cpp
class Solution {
public:
    long long gridGame(vector<vector<int>>& grid) {
        if (grid.size() != 2) { return -1; }
        if (grid[0].size() <= 1) { return 0; }

        const size_t cols = grid[0].size();
        int64_t u = reduce(grid[0].cbegin() + 1, grid[0].cend(), 0LL);
        int64_t d = 0;
        int64_t res = u;

        for (size_t j = 1; j < cols; j += 1) {
            u -= grid[0][j];
            d += grid[1][j - 1];
            res = std::min(res, std::max(u, d));
        }

        return res;
    }
};

File: algorithms/2_medium/G/02017_grid_game/solution.rs
impl Solution {
    pub fn grid_game(grid: Vec<Vec<i32>>) -> i64 {
        assert!(grid.len() == 2);
        let cols = grid[0].len();
        if cols <= 1 { return 0; }

        let mut u: i64 = grid[0].iter().skip(1).map(|&x| x as i64).sum();
        let mut d = 0_i64;
        let mut res = u;
        for j in 1..cols {
            u -= grid[0][j] as i64;
            d += grid[1][j - 1] as i64;
            res = res.min(u.max(d));
        }

        res
    }
}
